### PR TITLE
[7.x] Same-session ID request concurrency limiting

### PR DESCRIPTION
**Pull request by taylorotwell**
This PR implements a long requested feature (since basically the beginning of Laravel) for blocking concurrency requests from the same session until previous requests have finished. The basic use case being prevention of concurrent requests writing conflicting data to the session and data being lost.

There are many issues and StackOverflow / Laracasts / etc. threads about this but here is the most recent and the only one that still remains unclosed: https://github.com/laravel/framework/issues/18187

Implementing this doesn't seem particularly hard since Laravel offers atomic lock functionality out the box. If the `session.block` configuration variable is set to `true`, we attempt to obtain a lock on the a cache key matching `session:{session-id}`. Once we obtain one we process the request through the middleware and release the lock. If `session.block` is `false` there is no behavior change in session handling.

Route level blocking is implemented using a new `block` method on individual routes: `Route::post(...)->block()`

### This is a proof of concept and there are several things that need to be discussed:

It would be nice to ship locking support for the database cache driver. This seems possible (see: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/PdoStore.php). Offering this would open this feature up to a wider range of developers that may not have access to the currently supported `memcached`, `redis`, or `dynamodb` drivers.

### Caveats

This feature will not work with the `cookie` session driver.

Must be using `memcached`, `redis`, or `dynamodb` as a cache. As noted, would like to add `database` lock support.

### Remaining items summary:

- [x] Configurable block time at global and possibly route level
- [x] Ability to opt-in to blocking at route level only
- [x] Configurable cache driver to be used for session blocking
- [x] Database lock support

### CODE REVIEW

**Comment by trevorgehman** [comment]
I really appreciate your work on this. I think opt-in session locking is great. It seems like the simplest way to solve the issue.

Also the ability to block session writing on a route-level seems like a good idea. Personally, we have only a couple routes that actually write new data to the session (like login/switch accounts/etc.). However, if you block session writing on most routes won't you also have to still extend the session in those requests?

Have you reconsidered [this change](https://github.com/laravel/framework/pull/29410) which won't solve but will mitigate the issue for those who don't want to use session locking? You could implement [your idea](https://github.com/laravel/framework/issues/18187#issuecomment-622086329) to extend the session manually which solves the [issue of sessions expiring](https://github.com/laravel/framework/pull/29410#issuecomment-570284061) when you don't write them on every request. Seems like that ability will be needed anyways if you decide to move forward with a route-level `session.block` option.

**Comment by taylorotwell** [comment]
@trevorgehman every request will still extend the session lifetime as normal, so I don't quite follow the session extension stuff.

The general flow is:

- Request enters application
- Read session ID from cookie
- Attempt to acquire cache lock on session ID (if can't obtain it, wait for 10 seconds, if still can't obtain lock, throw exception)
- Execute request
- Write session
- Release lock

Session time extension still works the same.

**Review by usmanfourspan — APPROVED** [review]

**Comment by AbdelElrafa** [comment]
I'm just wondering if this might be possible to handle in a middleware, so that it could be applied to specific routes. Or would that be handled by `Route::get(...)->block($waitSeconds = 10)`?

**Comment by taylorotwell** [comment]
That would be handled by a new `block` method on the route, which I am writing as we speak.

File: src/Illuminate/Session/Middleware/StartSession.php [hunk 1]
```diff
@@ -44,11 +47,53 @@ public function handle($request, Closure $next)
             return $next($request);
         }
 
+        $session = $this->getSession($request);
+
+        if ($this->manager->shouldBlock()) {
+            return $this->handleRequestWhileBlocking($request, $session, $next);
+        } else {
```
Comment (antonkomarev) [inline src/Illuminate/Session/Middleware/StartSession.php:54]: `else` could be removed

**Comment by taylorotwell** [comment]
Route level blocking has been implemented.

File: src/Illuminate/Session/Middleware/StartSession.php (hunk 1 above)
Comment (GrahamCampbell) [reply src/Illuminate/Session/Middleware/StartSession.php:55]: Laravel usually keeps them for readability.

**Comment by taylorotwell** [comment]
Database lock support: https://github.com/laravel/framework/pull/32639

**Pull request merged**